feat(motion_velocity_planner): add new run_out module (#10388)
autowarefoundation/autoware_universe: planning/motion_velocity_planner/autoware_motion_velocity_run_out_module/README.md

# Run Out

## Role

The `run_out` module adds deceleration and stop points to the ego trajectory in order to prevent collisions with objects that are moving towards the ego vehicle path.

## Activation

This module is activated if the launch parameter `launch_mvp_run_out_module` is set to true.

## Inner-workings / Algorithms

> This module calculates the times when the ego vehicle and the objects are predicted to overlaps each other's trajectories.

Reviewer: ```suggestion
This module calculates the times when the ego vehicle and the objects are predicted to overlap each other's trajectories.
```

These geometries can be visualized on the debug markers with the `filtering_data_(ignore_objects|ignore_collisions|cut_predicted_paths)` namespaces.
The classification label corresponding to the published debug markers can be selected with parameter `debug.object_label`.

### 3. Dynamic objects filtering

In this step, objects and their predicted paths are filtered based on its classification label and the corresponding parameters `objects.{CLASSIFICATION_LABEL}`.

An object is ignored if one of the following condition is true:

- its classification label is not in the list defined by the `objects.target_labels` parameter;
- its velocity is bellow the `ignore.stopped_velocity_threshold` and `ignore.if_stopped` is set to `true`;
- its current footprint is inside one of the polygons prepared in the previous step.

However, if it was decided to stop for the object in the previous iteration,
or if a collision was detected with the object, then it cannot be ignored.

If an object is not ignored, its predicted path footprints are generated similarly to the ego footprint
First, we only keep predicted paths that have a confidence value above the `confidence_filtering.threshold` parameter.

> If, `confidence_filtering.only_use_highest` is set to `true` then for each object only the predicted paths that have the higher confidence value are kept.

Reviewer: ```suggestion
If, `confidence_filtering.only_use_highest` is set to `true` then for each object only the predicted path with the highest confidence value is kept.
```

Author: The plural is correct here as there are cases where multiple predicted paths have the same highest confidence value (e.g., 3 paths with `0.333...` confidence).
